### PR TITLE
fix(memory): spawnCapture returns stdout/stderr/exitCode; truncate parse-failure output (fixes #2151)

### PR DESCRIPTION
**Pull request by theshadow27**
## Summary
- `spawnCapture` now returns `{ stdout, stderr, exitCode }` instead of `string | null`, so `callHaiku` can distinguish subprocess failures (bad model, auth error, timeout) from parse failures — logs the actual exit code and stderr instead of a generic message
- Parse failures now show only first/last 50 chars of the response plus a path to a temp file with the full content, preventing 50KB+ stderr dumps on large repos
- Added `writeTempFile` to `MemoryDeps` for testability; `fetchClosedIssues` updated to use the new return type

## Test plan
- [x] New tests: exitCode-based failure logging with correct error message, no conflation of subprocess failures with parse failures, stderr output stays under 500 chars for 10KB bad response, gh non-zero exit handled gracefully, gh unavailable path covered
- [x] All 7449 existing tests pass
- [x] `bun typecheck` passes
- [x] `bun lint` passes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/command/src/commands/memory.ts [hunk 1]
```diff
@@ -271,11 +284,16 @@ export function defaultDeps(): MemoryDeps {
...
+      return {
+        stdout: r.stdout.toString(),
+        stderr: r.stderr.toString(),
+        exitCode: r.exitCode ?? 1,
+      };
+    },
+    writeTempFile: (content) => {
+      const path = join(tmpdir(), `memory-audit-response-${Date.now()}.txt`);
+      writeFileSync(path, content, "utf-8");
+      return path;
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/command/src/commands/memory.ts:301]: Nit — project is Unix-only one-shot CLI; macOS tmpdir is user-scoped so world-readable risk is minimal; concurrent audit runs are not a realistic scenario.

**Comment by theshadow27** [comment]
## QA Verification

**PR:** #2164 — fix(memory): spawnCapture returns stdout/stderr/exitCode; truncate parse-failure output (fixes #2151)
**Linked issue:** #2151

### Implementation Evidence
- `packages/command/src/commands/memory.ts:44` — `MemoryDeps.spawnCapture` now returns `{ stdout, stderr, exitCode }` (was `string | null`)
- `packages/command/src/commands/memory.ts:133-145` — `callHaiku` uses `result.exitCode !== 0` for subprocess failure, logs actual exit code + stderr; distinguishes from parse failure
- `packages/command/src/commands/memory.ts:151-169` — `fetchClosedIssues` uses `result.exitCode !== 0` for gh failure
- `packages/command/src/commands/memory.ts:44,46` — `writeTempFile` added to `MemoryDeps` for testability
- `packages/command/src/commands/memory.ts:223-233` — parse failure shows head/tail preview (≤100 chars total) + temp file path; prevents 50KB+ stderr dumps
- `packages/command/src/commands/memory.ts:281-297` — `defaultDeps` implementations for both `spawnCapture` (new return shape) and `writeTempFile`

### Test Coverage
- `packages/command/src/commands/memory.spec.ts:199-209` — non-zero exit code logged with stderr content (no "no response" conflation)
- `packages/command/src/commands/memory.spec.ts:211-223` — stderr from failed subprocess does NOT trigger parse-failure path
- `packages/command/src/commands/memory.spec.ts:235-248` — parse failure shows truncated preview + temp file path
- `packages/command/src/commands/memory.spec.ts:250-262` — parse failure with 10KB response stays under 500 chars of stderr
- `packages/command/src/commands/memory.spec.ts:264-279` — malformed gh JSON handled gracefully
- `packages/command/src/commands/memory.spec.ts:281-295` — gh unavailable (exitCode 127) handled gracefully
- 38 tests in memory.spec.ts, all pass

### Test Results
7449 pass, 22 todo, 0 fail across 303 files.

### CI Status
All checks green: build, check, coverage, detect, pty-test.

### Copilot Thread Disposition
- Thread #3271270487 (line 221) — **UNREPLIED / ACTIONABLE**: Empty stdout with exitCode 0 silently exits with code 1 and no diagnostic. The check `if (!rawResponse)` after `callHaiku` exits without logging why (no "no response from claude" message), making silent 0-exit/empty-stdout failures indistinguishable from a successful null return.
- Thread #3271270523 (line 231) — **UNREPLIED / ACTIONABLE**: `deps.writeTempFile(rawResponse)` is unguarded. If `writeFileSync` throws (disk full, permissions), the exception propagates and masks the original parse error — user sees a crash instead of the parse-failure message. Fix: wrap the `writeTempFile` call in try/catch and fall back to logging the preview-only path.
- Thread #3271270541 (line 296) — dismissed (nit; Unix-only one-shot CLI, macOS tmpdir is user-scoped, concurrent invocations not realistic).

### Remaining Work (non-blocking)
None from issue #2151.

### Outcome
**qa:fail** — 2 unreplied actionable Copilot threads block merge per the Step 5b gate (UNREPLIED_COUNT=2). Action for implementation:
1. Add a diagnostic message before `deps.exit(1)` when `callHaiku` returns an empty string (`!rawResponse` but truthy `rawResponse === ""`), e.g.: `deps.logError("memory audit: claude returned empty response")`.
2. Wrap `deps.writeTempFile(rawResponse)` in a try/catch in `runMemoryAudit`; on failure, skip the temp file hint and log the preview-only path.

File: packages/command/src/commands/memory.ts [hunk 2]
```diff
@@ -207,15 +217,18 @@ export async function runMemoryAudit(opts: { json: boolean }, deps: MemoryDeps):
   const rawResponse = callHaiku(prompt, deps);
 
   if (!rawResponse) {
-    deps.logError("memory audit: no response from Haiku");
     deps.exit(1);
   }
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/command/src/commands/memory.ts:222]: Fixed in follow-up commit 2033571: adds 'claude returned empty response' diagnostic when exitCode 0 but stdout is empty.

File: packages/command/src/commands/memory.ts [hunk 3]
```diff
@@ -207,15 +217,18 @@ export async function runMemoryAudit(opts: { json: boolean }, deps: MemoryDeps):
...
-    deps.logError("memory audit: failed to parse Haiku response");
-    deps.logError("--- raw response ---");
-    deps.logError(rawResponse);
+    const head = rawResponse.slice(0, 50);
+    const tail = rawResponse.length > 100 ? rawResponse.slice(-50) : "";
+    const preview = tail ? `${head}…${tail}` : head;
+    const tmpPath = deps.writeTempFile(rawResponse);
+    deps.logError(`memory audit: failed to parse Haiku response (${rawResponse.length} chars)`);
+    deps.logError(`  preview: ${preview}`);
+    deps.logError(`  full response saved to: ${tmpPath}`);
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/command/src/commands/memory.ts]: Fixed in follow-up commit 2033571: writeTempFile is now wrapped in try/catch to guard against disk-full/permission errors.

**Comment by theshadow27** [comment]
## QA Verification

**PR:** #2164 — fix(memory): spawnCapture returns stdout/stderr/exitCode; truncate parse-failure output (fixes #2151)
**Linked issue:** #2151

### Implementation Evidence
- `packages/command/src/commands/memory.ts:44` — `MemoryDeps.spawnCapture` signature changed from `() => string | null` to `() => { stdout, stderr, exitCode }`, plus new `writeTempFile` method added to deps
- `packages/command/src/commands/memory.ts:130-155` — `callHaiku` now checks `exitCode !== 0` and logs `claude exited with code N: <stderr>` for subprocess failures; stdout returned on success
- `packages/command/src/commands/memory.ts:160-170` — `fetchClosedIssues` updated to use `result.exitCode !== 0` and `result.stdout` from new return type
- `packages/command/src/commands/memory.ts:217-240` — Parse failure path: shows first/last 50 chars as preview, saves full response to temp file via `writeTempFile` wrapped in try/catch; empty response (exitCode 0, empty stdout) logs distinct "claude returned empty response" diagnostic
- `packages/command/src/commands/memory.ts:289-304` — `defaultDeps.spawnCapture` and new `defaultDeps.writeTempFile` implementations

### Test Coverage
- `packages/command/src/commands/memory.spec.ts` — 123 lines added, covering:
  - `logs stderr from failed claude subprocess without conflating with parse failure` (exitCode != 0 path, verifies error contains exit code)
  - `exits when Haiku returns unparseable response` (parse failure path)
  - `parse failure shows truncated preview and temp file path` (≤500 chars check + temp path in stderr)
  - `parse failure does not dump full response to stderr` (10KB bad response, stderr stays short)
  - `gh unavailable (non-zero exit) is handled gracefully`
  - `handles malformed gh JSON gracefully`
  - `logs diagnostic when claude returns empty stdout with exitCode 0`
  - `parse failure survives writeTempFile throwing`

### Test Results
CI `check` job passed (2m37s) — 7449 tests green. `coverage` job passed. `build` job passed. `pty-test` passed.

### CI Status
All 5 checks pass: https://github.com/theshadow27/mcp-cli/actions/runs/26142099375

### Copilot Threads
All 3 Copilot inline threads addressed:
- Thread on empty-response path: dismissed — fixed in commit 2033571
- Thread on writeTempFile missing try/catch: dismissed — fixed in commit 2033571
- Thread on temp filename collision/permissions: dismissed by owner (Unix-only, macOS tmpdir is user-scoped)

### Remaining Work (non-blocking)
None identified.

### Outcome
**qa:pass** — Both fix shapes from the issue are implemented and verified. `spawnCapture` returns `{ stdout, stderr, exitCode }`, subprocess failures log the actual exit code and stderr, parse failures show a truncated preview plus temp file path. Tests cover all new paths including edge cases. CI green.
